src/metrics_classification.rs

```diff
@@ -498,7 +489,17 @@ impl<D: Data<Elem = Pr>> BinaryClassification<&[bool]> for ArrayBase<D, Ix1> {
     }
 
     fn log_loss(&self, y: &[bool]) -> Result<f32> {
-        self.as_slice().unwrap().log_loss(y)
+        let clipped_probs: Vec<_> = self
+            .iter()
+            .map(|v| (*v).clamp(f32::EPSILON, 1. - f32::EPSILON))
+            .collect();
+        clipped_probs
+            .iter()
+            .zip(y.iter())
+            .map(|(a, b)| if *b { -a.ln() } else { -(1. - a).ln() })
+            .collect::<Array1<f32>>()
+            .mean()
+            .ok_or(Error::NotEnoughSamples)
```

Comment:
To avoid creating arrays. do:
```rust
let sum = self
  .iter()
  .map(|v| ...)
  .zip(y.iter())
  .map(|(a, b)| ...)
  .sum();
let len = self.len();
if len == 0 {
  Err(Error::NotEnoughSamples)
} else {
  Ok(sum / len)
}
```